___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

# WELCOME!

## Introduction
Welcome to "***AutoScout Data Analysis Project***". This is the capstone project of ***Data Analysis*** Module. **Auto Scout** data which using for this project, scraped from the on-line car trading company in 2019, contains many features of 9 different car models. In this project, you will have the opportunity to apply many commonly used algorithms for Data Cleaning and Exploratory Data Analysis by using many Python libraries such as Numpy, Pandas, Matplotlib, Seaborn, Scipy you will analyze clean dataset.

The project consists of 3 parts:
* First part is related with 'data cleaning'. It deals with Incorrect Headers, Incorrect Format, Anomalies, Dropping useless columns.
* Second part is related with 'filling data'. It deals with Missing Values. Categorical to numeric transformation is done.
* Third part is related with 'handling outliers of data' via Visualisation libraries. Some insights are extracted.


# PART- 1 `( Missing Values )`

In [1]:
import pandas as pd
import numpy as np

#Reading the data 
df0 = pd.read_json ('scout_car.json', lines = True)
df = df0
df.sample(10)

url      make_model  \
13316  https://www.autoscout24.com//offers/renault-cl...    Renault Clio   
7696   https://www.autoscout24.com//offers/opel-astra...      Opel Astra   
14368  https://www.autoscout24.com//offers/renault-cl...    Renault Clio   
1554   https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
15709  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
11443  https://www.autoscout24.com//offers/opel-insig...   Opel Insignia   
2975   https://www.autoscout24.com//offers/audi-a3-sp...         Audi A3   
9578   https://www.autoscout24.com//offers/opel-corsa...      Opel Corsa   
8102   https://www.autoscout24.com//offers/opel-astra...      Opel Astra   
10390  https://www.autoscout24.com//offers/opel-corsa...      Opel Corsa   

                                       short_description      body_type  \
13316  Energy TCe 120 EDC,Automatik,PDC,Nav,Klimaau.,...         Sedans   
7696   K INNOVATION 1.4 Turbo AT EU6d-T Voll LED Navi...        Compact   
14368    1.6 T 220 EDC 18 RS CUIR CAMERA MONITOR SUPERBE         Sedans   
1554        Sportback sport 1.4 TDI NAV XEN+ PDC SHZ GRA         Sedans   
15709                              TCe 225 Energy Intens            Van   
11443  ST.2,0l Innovation AHK abn. SHZ PDC Rückfahrka...  Station wagon   
2975               Sportback 1.4 TFSI Attraction S-T 125        Compact   
9578                           1.2 5 porte Black Edition         Sedans   
8102          1.0 Turbo 120 Jahre *Kamera*LED*Winterpak*        Compact   
10390                                        1.2 Edition        Compact   

       price             vat          km registration         prev_owner  kW  \
13316  11990            None   17,890 km      11/2016   1 previous owner NaN   
7696   26899  VAT deductible      100 km      06/2019               None NaN   
14368  25990            None    7,900 km      07/2018   1 previous owner NaN   
1554   16140  VAT deductible   13,800 km      08/2018   1 previous owner NaN   
15709  37490  VAT deductible    1,150 km      11/2018   1 previous owner NaN   
11443  19499  VAT deductible   37,125 km      03/2017  3 previous owners NaN   
2975   13950            None  118,032 km      06/2016               None NaN   
9578    9390  VAT deductible   43,000 km      03/2018               None NaN   
8102   18979  VAT deductible      500 km      05/2019   1 previous owner NaN   
10390  11250  VAT deductible      100 km      02/2019   1 previous owner NaN   

       ...                                        description  \
13316  ...                                           [\n, \n]   
7696   ...  [\n, Getriebe:,  Automatik, Technik:,  Bordcom...   
14368  ...  [\nAUTO PASSION 18 SPECIALISTE DE L'AUTOMOBILE...   
1554   ...  [\n, Technik & Sicherheit:, Klimaautomatik, Xe...   
15709  ...  [\n, Audio-Navigationssystem R-Link 2 mit Touc...   
11443  ...  [\n, Getriebe:,  Automatik, Technik:,  Bordcom...   
2975   ...  [\nEn Traverauto (Desde 1978,Opel desde 1986),...   
9578   ...  [\n, VEICOLO IN PROMOZIONE, •Vettura finanziab...   
8102   ...  [\nIhre Ansprechpartner für dieses KFZ:,  ,  ,...   
10390  ...  [\n,  * Lackierung: Uni Sonderfarbe,  * Außens...   

        Emission Label    Gears Country version Electricity consumption  \
13316              NaN  [\n6\n]   [\nGermany\n]                     NaN   
7696   [\n4 (Green)\n]  [\n6\n]   [\nGermany\n]                     NaN   
14368              NaN  [\n6\n]    [\nFrance\n]                     NaN   
1554               NaN      NaN   [\nGermany\n]                     NaN   
15709  [\n4 (Green)\n]      NaN             NaN                     NaN   
11443              NaN  [\n6\n]             NaN                     NaN   
2975               NaN  [\n7\n]             NaN                     NaN   
9578               NaN  [\n5\n]             NaN                     NaN   
8102               NaN  [\n5\n]   [\nGermany\n]                     NaN   
10390              NaN  [\n5\n]   [\nGerman

In [2]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [3]:
object_col = df.select_dtypes(include = "object").columns 

for col in object_col:
  print(col)
  print("--"*5)
  print(df[col].value_counts(dropna = False))
  print( "--"*10)

url
----------
https://www.autoscout24.com//offers/opel-astra-1-6-cdti-active-euro-6-diesel-red-2d12f60d-5888-48d9-9e3c-e976660b0ec0                                     1
https://www.autoscout24.com//offers/audi-a1-sportback-1-0tfsi-klimaautom-mmi-radio-sitzh-alu-gasoline-white-175f28d3-cfd3-4130-83fa-704855249b58          1
https://www.autoscout24.com//offers/opel-astra-k-sports-tourer-on-start-stop-1-4-turbo-navi-rueckf-gasoline-black-9b8a482e-5f6e-47e4-ab2b-3229caeb827b    1
https://www.autoscout24.com//offers/audi-a3-sportback-basis-xenon-pdc-mmi-sitzheiz-ssa-gasoline-silver-6b1188f6-9b12-4690-a78a-3ccf3bd0a122               1
https://www.autoscout24.com//offers/audi-a3-sportback-2-0-tdi-dsg-navi-xenon-pdc-diesel-black-07db7761-8988-4fec-8925-0349d5aca7de                        1
                                                                                                                                                         ..
https://www.autoscout24.com//offers/audi-a1-sport

NaN                3175
[\nXT61649\n]        27
[\nHM53619\n]        27
[\nTP62881\n]        27
[\nJV03654\n]        27
                   ... 
[\n9277941\n]         1
[\n19155-1\n]         1
[\n3124254\n]         1
[\nJ1109546\n]        1
[\n040394-MO\n]       1
Name: Offer Number, Length: 11441, dtype: int64
--------------------
First Registration
----------
[\n, 2018, \n]    4522
[\n, 2016, \n]    3674
[\n, 2017, \n]    3273
[\n, 2019, \n]    2853
NaN               1597
Name: First Registration, dtype: int64
--------------------
Body Color
----------
[\n, Black, \n]     3745
[\n, Grey, \n]      3505
[\n, White, \n]     3406
[\n, Silver, \n]    1647
[\n, Blue, \n]      1431
[\n, Red, \n]        957
NaN                  597
[\n, Brown, \n]      289
[\n, Green, \n]      154
[\n, Beige, \n]      108
[\n, Yellow, \n]      51
[\n, Violet, \n]      18
[\n, Bronze, \n]       6
[\n, Orange, \n]       3
[\n, Gold, \n]         2
Name: Body Color, dtype: int64
--------------------
Paint Type
--

NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                     920
[Air conditioning, Electrical side mirrors, Hill Holder, Power windows]                                                                                                                                                                                                                                                                                                                                                                                 216
[Air conditioning, Electrical side mirrors, Power windows]                                                      

### Examining Each Column

### 1 . URL

In [4]:
#df['url'].value_counts(dropna = False)

In [5]:
df.drop("url",axis=1,inplace=True)

<font color='light blue'>
a. Column 'url' was used for web scrabing and gives information for each product. <br>
b. Dropped, not related to target field.
</font>

### 2. make_model

In [6]:
df['make_model'].value_counts(dropna = False)

Audi A3           3097
Audi A1           2614
Opel Insignia     2598
Opel Astra        2526
Opel Corsa        2219
Renault Clio      1839
Renault Espace     991
Renault Duster      34
Audi A2              1
Name: make_model, dtype: int64

In [7]:
df.drop(["Make","Model"],axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'make_model' and to dos. </font>
<font color='light blue'>
    
a. This is a categorical column <br>
b. The columns icludes the name of the make and model of the cars.<br>
c. We have 3 makes and 9 models <br>
d. All the values are strings  <br>
e. Clean data, no need to work on <br>
f. Directly related to the target value<br>
g. Since "Make_Model" and "Model-Make Columns" are equel, we dropped these columns
</font>

### 3. short_description

In [8]:
sd_disp = df['short_description'].str.findall('\d\.\d').str[0].astype(float)
df['sd_disp'] = sd_disp*1000
df['sd_disp'] = df['sd_disp'].replace(1600,1598).replace(1800,1798)

In [9]:
df.loc[df['sd_disp']<800,'sd_disp'] = np.nan

<font color='orange'> ** Impressions on the column 'short_description' and to dos. </font>
<font color='light blue'>
    
a. This is a categorical column <br>
b. The column icludes some details about the cars as free-text.<br>
c. The data of dispalement extracted to use the NaN's of Displacement column.  <br>
d. SInce it is not consistent  for each model, it should be dropped<br>
</font>

### 4. body_type

In [10]:
df.body_type.value_counts(dropna = False)

Sedans           7903
Station wagon    3553
Compact          3153
Van               783
Other             290
Transporter        88
NaN                60
Off-Road           56
Coupe              25
Convertible         8
Name: body_type, dtype: int64

In [11]:
#60 NaNs filled with 'Other'
df['body_type'] = df['body_type'].fillna('Other')

In [12]:
df.body_type.value_counts(dropna = False)

Sedans           7903
Station wagon    3553
Compact          3153
Van               783
Other             350
Transporter        88
Off-Road           56
Coupe              25
Convertible         8
Name: body_type, dtype: int64

<font color='orange'> ** Impressions on the column 'body_type' and to dos. </font>
<font color='light blue'>
    
a. This is a categorical column <br>
b. The column icludes body_types of the cars.<br>
s. There are 60 NaNs, they could be filled as 'Other'  <br>
d. This data should be checked with other features <br>
e. Directly related to the target value but need to think on that
</font>

### 5. Price

In [13]:
df.price.value_counts(dropna=False)


14990    154
15990    151
10990    139
15900    106
17990    102
        ... 
31940      1
15540      1
21649      1
27790      1
34815      1
Name: price, Length: 2956, dtype: int64

<font color='orange'> ** Impressions on the column 'price' </font>
<font color='light blue'>
    
a. This is target value <br>
b. The type is integer and data is clean
</font>

### 6. Vat

In [14]:
df.vat.value_counts(dropna=False)

VAT deductible      10980
NaN                  4513
Price negotiable      426
Name: vat, dtype: int64

In [15]:
#4513 NaNs filled with 'VAT not deductible'
df['vat'] = df['vat'].fillna('VAT not deductible')

In [16]:
df.vat.value_counts(dropna=False)

VAT deductible        10980
VAT not deductible     4513
Price negotiable        426
Name: vat, dtype: int64

<font color='orange'> ** Impressions on the column 'vat' </font>
<font color='light blue'>
    
a. This column is related to target value <br>
b. The options other than 'Vat Deductible' could be 0 and 'vat deductible' as 1
</font>

### 7. km

In [17]:
df['km']=df['km'].str.replace(',' , '').str.findall('\d+').str[0].astype(float)
df['km'].value_counts(dropna = False)

10.0        1045
NaN         1024
1.0          367
5.0          170
50.0         148
            ... 
160542.0       1
20719.0        1
91910.0        1
39860.0        1
57889.0        1
Name: km, Length: 6690, dtype: int64

In [18]:
#???
df['km'] = df['km'].fillna(df.groupby(['make_model','registration'])['km'].transform('mean'))

In [19]:
df['km'].value_counts(dropna = False)

10.000000        1045
1.000000          367
705.505155        229
793.189781        223
5.000000          170
                 ... 
18636.000000        1
94220.000000        1
115900.000000       1
26217.000000        1
57889.000000        1
Name: km, Length: 6753, dtype: int64

<font color='orange'> ** Impressions on the column 'km'  </font>
<font color='light blue'>
    
a. This should be a numerical column and type should be float, DONE <br>
b. The column icludes km of the cars.<br>
e. Directly related to the target value
</font>

### 8. Registration

In [20]:
"""
from datetime import datetime
def cleaning_registration(a):
    if a == '-/-':
        return np.nan
    else:
        return '01/'+a
    
def difference_in_months(a):
    if pd.isna(a):
        return a
    else:
        b = datetime.strptime('01/12/2019', '%d/%m/%Y').date()
        a = datetime.strptime(a, '%d/%m/%Y').date()
        return (b.year - a.year) * 12 + (b.month - a.month)


df['age_in_months'] = df.registration.apply(cleaning_registration)
df['age_in_months'] = df['age_in_months'].apply(difference_in_months)
df['age_in_months'].value_counts(dropna = False)
"""

"\nfrom datetime import datetime\ndef cleaning_registration(a):\n    if a == '-/-':\n        return np.nan\n    else:\n        return '01/'+a\n    \ndef difference_in_months(a):\n    if pd.isna(a):\n        return a\n    else:\n        b = datetime.strptime('01/12/2019', '%d/%m/%Y').date()\n        a = datetime.strptime(a, '%d/%m/%Y').date()\n        return (b.year - a.year) * 12 + (b.month - a.month)\n\n\ndf['age_in_months'] = df.registration.apply(cleaning_registration)\ndf['age_in_months'] = df['age_in_months'].apply(difference_in_months)\ndf['age_in_months'].value_counts(dropna = False)\n"

In [21]:
df["registration"]=df["registration"].replace("-/-",np.NaN)

In [22]:
df["registration"]=pd.to_datetime(df["registration"]).dt.year

In [23]:
df["registration"].value_counts(dropna=False)

2018.0    4522
2016.0    3674
2017.0    3273
2019.0    2853
NaN       1597
Name: registration, dtype: int64

In [24]:
df.groupby('Type')["registration"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

In [ ]:
df.groupby('Type')['registration'].apply(lambda x:x.mode()[0])

In [ ]:
#???
df['registration'] = df['registration'].fillna(df.groupby('Type')['registration'].apply(lambda x:x.mode()[0]))

In [ ]:
df["registration"].value_counts(dropna=False)

<font color='orange'> ** Impressions on the column 'registration' </font>
<font color='light blue'>
    
a. This should be a numerical column and type should be float <br>
b. The column icludes registration dates of the cars.<br>
c. There are 1597 '-/-' . The NaNs could be checked with Type1: " Demonstration and new" could be filled as 0 <br>
d. We could calculate the age as year or months. DONE As Year <br>
e. Directly related to the target value
</font>

### 9. prev_owner

In [ ]:
df['prev_owner']=df['prev_owner'].str.findall("\d+").str[0].astype(float)

In [ ]:
df['prev_owner'].value_counts(dropna = False)

In [ ]:
df["Previous Owners"]=df["Previous Owners"].str.strip().astype(float)

In [ ]:
df[["prev_owner","Previous Owners"]]

In [ ]:
def prev_owner_combine(p1,p2):
    if p1 == p2:
        return p1
    elif np.isnan(p1) :
        if np.isnan(p2):
            return np.nan
        else:
            return p2
    elif np.isnan(p2):
        if np.isnan(p1):
            return np.nan
        else:
            return p1
    else:
        return 'conflict'

In [ ]:
df["prev_owner"]=df.apply(lambda x: prev_owner_combine(x['prev_owner'],x['Previous Owners']), axis=1)

In [ ]:
df.drop('Previous Owners',axis=1,inplace=True)

In [ ]:
df['prev_owner'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'prev_owner' </font>
<font color='light blue'>
    
a. This should be a numerical column and type should be int <br>
b. The column icludes the the number of previous owners of the cars.<br>
c. There are 6828 'NaN' and %43 of total values, but after comparing with previou owne volumn. They could be filled <br>
d. We could check if the NaN values are 0 or not with checking the registration dates <br>
e. Relation to the target value could be checked with correlation
</font>

### 10. kW

In [ ]:
df.kW.value_counts(dropna = False)
df.drop("kW",axis=1,inplace=True)

<font color='light blue'>
    
a. All the values are NaN <br>
b. Dropped.<br>

</font>

### 11.hp

In [ ]:
df['hp']=df['hp'].map(lambda x: x.rstrip(' kW')).replace("-",np.NaN).astype("float")

<font color='orange'> ** Impressions on the column 'hp' </font>
<font color='light blue'>
    
a. This should be a numerical column and type should be float, DONE <br>
b. The column icludes power of the cars.<br>
c. There are  88'-' .Replaced with np.nan   <br>
d. 'kW' should be removed. DONE <br>
e. Relation to the target value would be checked with correlation
</font>

### 12. Type

In [ ]:
df.Type=df.Type.str[1].str.strip()

In [ ]:
df.Type.value_counts(dropna = False)

In [ ]:
df['Type'] = df['Type'].fillna('Used')

In [ ]:
df.Type.value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Type' </font>
<font color='light blue'>
    
a. This column includes list of lists about two different data <br>
b. first part is Used/new/.. and second part is about fuel.<br>
c. Since the fuel data is same with the column fuel. This part is neglegted.   <br>
 <br>
</font>

### 13.Previous Owners


<font color='light blue'>
    
a. Deleted after combined with prev_owners columns
</font>

### 14.Next Inspection

In [ ]:
listINS=[]
for i in df["Next Inspection"]:
    if type(i)==float:
        listINS.append(i)
    elif type(i)==list:
        listINS.append(i[0].strip())
    else:
        listINS.append(i.replace("\n",""))

In [ ]:
df["Next Inspection"]=listINS


In [ ]:
df["Next Inspection"]=pd.to_datetime(df["Next Inspection"])

In [ ]:
df["Next Inspection"] = df["Next Inspection"].fillna('2019-12-01 00:00:00')

In [ ]:
df["Next Inspection"].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Next Inspection' </font>
<font color='light blue'>
    
a. This column includes many '\n' symbols and they should be cleaned, DONE <br>
b. The column icludes date of next inspection.<br>
</font>

### 15. Inspection new

In [ ]:
def inspection_new(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    elif type(a)== str:
        return a.replace('\n', '')
    else:
        return a
    

df['Inspection new'] = df['Inspection new'].apply(inspection_new)
df['Inspection new'].value_counts(dropna = False, normalize= True)

In [ ]:
df['Inspection new'] = df['Inspection new'].fillna('No')

In [ ]:
df['Inspection new'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Inspection New' </font>
<font color='light blue'>
    
a. This column includes the boolean data if the inspection is new or not, DONE <br>
b. The column icludes date of next inspection.<br>
c. The percentage of NaN is 75% and they could be filled with No which means it is not a new inspection   <br>
d. Relation to the target value would be checked with correlation
</font>

### 16-Warranty

In [ ]:
df['Warranty']

In [ ]:
import re
def clean_alt_list(a):
    if type(a) == list:
        b = re.findall(r'\d+', a[0])
        if len(b)== 0:
            return np.nan
        else:
            return b[0]
    elif type(a) ==str:
        b = re.findall(r'\d+', a)
        if len(b)== 0:
            return np.nan
        else:
            return b[0]
    else:
        return a

df['Warranty'] = df['Warranty'].apply(clean_alt_list)
df = df.rename({'Warranty': 'Warranty(months)'}, axis=1)


In [ ]:
df['Warranty(months)'].value_counts(dropna = False)

In [ ]:
df['Warranty(months)'] = df['Warranty(months)'].fillna(0.0)

In [ ]:
df['Warranty(months)'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Warranty' </font>
<font color='light blue'>
    
a. This column includes many '\n' symbols and they should be cleaned, DONE <br>
b. The column icludes date of warranty in months, also some additional data.<br>
c. The percentage of NaN is 70% andand could be filled with 0 which meand there is no warranty   <br>
d. Relation to the target value would be checked with correlation <br>
e. The name of the column is change to Warranty(months)
</font>

### 17. Full Service

In [ ]:
df.drop('Full Service',axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Full Service' </font>
<font color='light blue'>
    
a. This column includes repetitive data and is dropped
</font>

### 18. 'Non-smoking Vehicle'

In [ ]:
df.drop('Non-smoking Vehicle',axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Non-smoking Vehicle' </font>
<font color='light blue'>
    
a. This column includes repetitive data and is dropped
</font>

### 19. 'null'

In [ ]:
df.drop('null',axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'null' </font>
<font color='light blue'>
    
a. This column includes empty lists and is dropped
</font>

### 20. MAKE

<font color='orange'> ** Impressions on the column 'Make' </font>
<font color='light blue'>
    
a. This column is repetitive and dropped</font>

### 21. 'Model'


<font color='orange'> ** Impressions on the column 'Model' </font>
<font color='light blue'>
    
a. This column is repetitive and dropped</font>

### 22. 'Offer Number'

In [ ]:
df.drop('Offer Number',axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Offer Number' </font>
<font color='light blue'>
    
a. This column includes offer numbers.<br>
c. It is not a related value to the target value and is dropped
</font>

### 23. 'First Registration'

In [ ]:
df.drop("First Registration",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'First Registration' </font>
<font color='light blue'>
    
a. This column includes First registration years.<br>
c. Checked with registration.since they have the same data, is is dropped.
</font>

### 24. 'Body Color'

In [ ]:
def cleaning_body_color(a):
    if type(a) == list:
        return a[1]
    else:
        return a
    
df['Body Color'] = df['Body Color'].apply(cleaning_body_color)

In [ ]:
df['Body Color'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Body Color' </font>
<font color='light blue'>
    
a. This column includes body colors.<br>
b. The numbers were within lists as second element. They are extracted <br>
c. There are 597 NaNs. The NaNs could be filled with 'Body Color Original'<br>
d. The relation with price could be checked with correlation
</font>

### 25. 'Paint Type'

In [ ]:
def cleaning_the_paint_type(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Paint Type'] = df['Paint Type'].apply(cleaning_the_paint_type)

In [ ]:
df['Paint Type'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Paint Type' </font>
<font color='light blue'>
    
a. This column includes paint types.<br>
b. The numbers were within lists. Extracted and newline symbols were cleaned<br>
c. Relation with prices should be checked. <br>
d. There are 5772 NaNs.
</font>

### 26. 'Body Color Original'

In [ ]:
def cleaning_original_color(a):
    if type(a) == list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Body Color Original'] = df['Body Color Original'].apply(cleaning_original_color)  

In [ ]:
df['Body Color Original'].value_counts(dropna= False)

In [ ]:
df.groupby('Body Color')['Body Color Original'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Body Color Original' </font>
<font color='light blue'>
    
a. This column includes original body color. It is more in detail and in German.<br>
b. This columns could be used to fill the NaNs in 'Body Color' column<br>
c. It could be dropped <br>
</font>

### 27. 'Upholstery'

In [ ]:
def cleaning_Upholstery(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Upholstery'] = df['Upholstery'].apply(cleaning_Upholstery) 

In [ ]:
df['Upholstery'].value_counts(dropna = False)

<font color='orange'> ** Impressions on the column 'Upholstery' </font>
<font color='light blue'>
    
a. This column includes Upholstery type and color. .<br>
b. There are 3720 NaNs, maybe they could be filled as other<br>
c. The relation should be checked with the target value <br>
</font>

### 28. 'Body'

In [ ]:
def cleaning_body(a):
    if type(a)== list:
        return a[1]
    else:
        return a
    
df['Body'] = df['Body'].apply(cleaning_body)

In [ ]:
df['Body'] = df['Body'].replace(np.nan, 'None')

In [ ]:
df.groupby('Body')['body_type'].value_counts(dropna = False)

In [ ]:
df.drop("Body",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Body' </font>
<font color='light blue'>
    
a. This column includes body type and color. .<br>
b. It is totally same with the column 'body_type'<br>
c. Since the data of body_type and body is same.This column dropped<br>
</font>

### 29. Nr. of Doors

In [ ]:
def cleaning_Doors(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Nr. of Doors'] = df['Nr. of Doors'].apply(cleaning_Doors) 

In [ ]:
df['Nr. of Doors'].value_counts(dropna = False)

In [224]:
df.columns

Index(['url', 'make_model', 'short_description', 'body_type', 'price', 'vat',
       'km', 'registration', 'prev_owner', 'kW', 'hp', 'Type',
       'Previous Owners', 'Next Inspection', 'Inspection new',
       'Warranty(months)', 'Full Service', 'Non-smoking Vehicle', 'null',
       'Make', 'Model', 'Offer Number', 'First Registration', 'Body Color',
       'Paint Type', 'Body Color Original', 'Upholstery', 'Body',
       'Nr. of Doors', 'Nr. of Seats', 'Model Code', 'Gearing Type',
       'Displacement', 'Cylinders', 'Weight', 'Drive chain', 'Fuel',
       'Consumption', 'CO2 Emission', 'Emission Class',
       '\nComfort & Convenience\n', '\nEntertainment & Media\n', '\nExtras\n',
       '\nSafety & Security\n', 'description', 'Emission Label', 'Gears',
       'Country version', 'Electricity consumption', 'Last Service Date',
       'Other Fuel Types', 'Availability', 'Last Timing Belt Service Date',
       'Available from'],
      dtype='object')

In [235]:
??
df.groupby(['body_type', 'make_model'])['Nr. of Doors'].apply(lambda x: x.mode())

body_type      make_model       
Compact        Audi A1         0    5
               Audi A3         0    5
               Opel Astra      0    5
               Opel Corsa      0    5
               Opel Insignia   0    5
               Renault Clio    0    5
               Renault Espace  0    5
Convertible    Audi A3         0    2
Coupe          Audi A1         0    2
                               1    5
               Audi A3         0    3
               Opel Astra      0    3
                               1    5
               Opel Corsa      0    3
               Opel Insignia   0    4
               Renault Clio    0    4
               Renault Espace  0    5
Off-Road       Audi A2         0    5
               Opel Astra      0    5
               Opel Corsa      0    5
               Opel Insignia   0    5
               Renault Clio    0    5
               Renault Duster  0    5
               Renault Espace  0    5
Other          Audi A1         0    5
               Au

<font color='orange'> ** Impressions on the column 'Nr. of Doors' </font>
<font color='light blue'>
    
a. This column includes number of doors. .<br>
b. There are 212 NaNs<br>
c. The NaNs could be filled with the mode of the body type or model<br>
</font>

### 30. 'Nr. of Seats'

In [56]:
def cleaning_Seats(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Nr. of Seats'] = df['Nr. of Seats'].apply(cleaning_Seats) 

In [57]:
df['Nr. of Seats'].value_counts(dropna = False)

5      13336
4       1125
NaN      977
7        362
2        116
6          2
3          1
Name: Nr. of Seats, dtype: int64

<font color='orange'> ** Impressions on the column 'Nr. of Seats' </font>
<font color='light blue'>
    
a. This column includes number of seats .<br>
b. There are 977 NaNs
</font>

### 31. 'Model Code'

In [58]:
def cleaning_model_code(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
    
df['Model Code'] = df['Model Code'].apply(cleaning_model_code) 

In [59]:
df['Model Code'].value_counts(dropna = False, normalize = True )

NaN         0.687292
0035/BCB    0.016835
0588/BNO    0.015390
0588/BDB    0.012941
0588/BHX    0.011810
              ...   
0588/BN0    0.000063
0588/AWS    0.000063
1234/123    0.000063
0035/BGY    0.000063
0000/ACP    0.000063
Name: Model Code, Length: 233, dtype: float64

<font color='orange'> ** Impressions on the column 'Model Code' </font>
<font color='light blue'>
    
a. This column includes model code of cars .<br>
b. Each model code have detailed meaning but should be checked in detail.
</font>

### 32. 'Gearing Type'

In [60]:
def cleaning_gearing(a):
    if type(a)== list:
        return a[1]
    else:
        return a
    
df['Gearing Type'] = df['Gearing Type'].apply(cleaning_gearing) 

In [61]:
df['Gearing Type'].value_counts(dropna = False)

Manual            8153
Automatic         7297
Semi-automatic     469
Name: Gearing Type, dtype: int64

<font color='orange'> ** Impressions on the column 'Gearing Type' </font>
<font color='light blue'>
    
a. This column includes gearing types of cars .<br>
b. The data is clean.
</font>

### 33. 'Displacement' 

In [62]:
df['Displacement']= df['Displacement'].str[0].str.replace(',','').str.findall('\d+').str[0]

In [63]:
#df["Displacement"]=df["Displacement"].str[0].str.replace("cc","").str.replace(",","").str.strip()
#df["Displacement"]=pd.to_numeric(df["Displacement"])

In [64]:
def disp(d1,d2):
    if (d1>4000) | (d1<700) | np.isnan(d1):
        if np.isnan(d2):
            return d1
        else:
            return d2
    else:
        return d1

In [65]:
df['Displacement'] = df.apply(lambda x: disp(x['Displacement'],x['sd_disp']),axis=1)

TypeError: '>' not supported between instances of 'str' and 'int'

In [66]:
df.drop(["short_description","sd_disp"],axis=1,inplace=True)

### 34. 'Cylinders'

In [67]:
def cleaning_cylinders(a):
    if type(a)== list:
        b = a[0].replace('\n', '')
        return b
    else:
        return a
df['Cylinders'] = df['Cylinders'].apply(cleaning_cylinders)

In [68]:
df['Cylinders'].value_counts(dropna = False, normalize = True)

4      0.509140
NaN    0.356806
3      0.132169
5      0.001382
6      0.000188
2      0.000126
8      0.000126
1      0.000063
Name: Cylinders, dtype: float64

** Impressions on the column 'Cylinders'

a. This column includes number of cylinders of cars .
b. 35% of the data is NaN.

### 35. 'Weight' 

In [69]:
df ['Weight'] = df['Weight'].str[0].str.replace(',','').str.findall('\d+').str[0]

In [70]:
df.rename(columns = {'Weight':'Weight(kg)'}, inplace = True)

In [71]:
df['Weight(kg)'].value_counts(dropna = False)

NaN     6974
1163     574
1360     356
1165     301
1335     242
        ... 
1761       1
1649       1
1223       1
102        1
1331       1
Name: Weight(kg), Length: 435, dtype: int64

### 35.1 Drive Chain

In [72]:
def cleaning_drive_chain(a):
    if type(a)== list:
        return a[0].replace('\n', '')
    else:
        return a
df['Drive chain'] = df['Drive chain'].apply(cleaning_drive_chain)    

In [73]:
df['Drive chain'].value_counts(dropna = False, normalize = True)

front    0.558201
NaN      0.430806
4WD      0.010742
rear     0.000251
Name: Drive chain, dtype: float64

<font color='orange'> ** Impressions on the column 'Drive Chain' </font>
<font color='light blue'>
    
a. This column includes drive-chain types of cars .<br>
b. The data is clean but 43% of them are NaN
</font>

### 36. ''Fuel''

In [74]:
df.Fuel = df.Fuel.apply(lambda x : x[1])

In [77]:
df['Fuel'].replace(['Diesel (Particulate Filter)','Diesel'], 'Diesel',inplace = True)

In [78]:
gasoline = ['Super 95',
            'Gasoline',
            'Super 95 / Regular/Benzine 91',
            'Regular/Benzine 91',
            'Super E10 95',
            'Super 95 (Particulate Filter)',
            'Super 95 / Super E10 95',
            'Regular/Benzine 91 (Particulate Filter)',
            'Super 95 / Super Plus 98',
            'Gasoline (Particulate Filter)',
            'Super 95 / Super Plus 98 / Super E10 95 / Super Plus E10 98',
            'Super E10 95 / Super 95',
            'Super 95 / Super E10 95 (Particulate Filter)',
            'Super E10 95 / Super 95 / Super Plus 98 / Super Plus E10 98',
            'Super 95 / Super Plus 98 (Particulate Filter)',
            'Regular/Benzine 91 / Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super E10 95 / Super Plus E10 98',
            'Regular/Benzine E10 91',
            'Super 95 / Super Plus 98 / Super E10 95',
            'Super Plus 98',
            'Super E10 95 (Particulate Filter)', 
            'Super 95 / Regular/Benzine 91 / Super Plus 98 / Regular/Benzine E10 91 / Super E10 95 / Super Plus E10 98',
            'Regular/Benzine 91 / Regular/Benzine E10 91',
            'Super 95 / Super Plus 98 / Super Plus E10 98 / Super E10 95',
            'Super E10 95 / Super Plus E10 98',
            'Super 95 / Super Plus 98 / Super E10 95 / Super Plus E10 98 (Particulate Filter)',
            'Regular/Benzine 91 / Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98 / Super E10 95',
            'Super 95 / Super E10 95 / Super Plus E10 98 / Super Plus 98',
            'Super E10 95 / Super 95 / Super Plus 98 / Super Plus E10 98 (Particulate Filter)',
            'Super 95 / Regular/Benzine 91 / Super Plus 98',
            'Super E10 95 / Regular/Benzine 91',
            'Regular/Benzine 91 / Super 95 / Super Plus 98 / Super Plus E10 98 / Super E10 95 / Regular/Benzine E10 91',
            'Super Plus 98 / Regular/Benzine 91',
            'Regular/Benzine E10 91 / Super E10 95 / Super Plus E10 98 / Super Plus 98 / Super 95 / Regular/Benzine 91',
            'Super Plus 98 / Super 95 / Super Plus E10 98',
            'Super 95 / Regular/Benzine E10 91',
            'Super E10 95 / Regular/Benzine 91 / Regular/Benzine E10 91 / Super 95',
            'Super 95 / Regular/Benzine 91 / Super E10 95 / Super Plus E10 98 / Super Plus 98',
            'Super 95 / Super Plus 98 / Super Plus E10 98',
            'Regular/Benzine 91 / Super 95 / Super Plus 98 / Super Plus E10 98 / Super E10 95 / Regular/Benzine E10 91 (Particulate Filter)',
            'Regular/Benzine 91 / Super 95 / Regular/Benzine E10 91',
            'Regular/Benzine E10 91 / Regular/Benzine 91 / Super 95 / Super Plus 98 / Super E10 95 / Super Plus E10 98',
            'Super 95 / Super E10 95 / Super Plus E10 98',
            'Super 95 / Regular/Benzine 91 / Super E10 95 / Super Plus 98 / Super Plus E10 98 / Regular/Benzine E10 91',
            'Regular/Benzine 91 / Super 95',
            'Regular/Benzine 91 / Super 95 / Regular/Benzine E10 91 / Super E10 95 / Super Plus 98 / Super Plus E10 98',
            'Super 95 / Regular/Benzine 91 (Particulate Filter)',
            'Regular/Benzine 91 / Super 95 / Super Plus E10 98 / Super E10 95 / Regular/Benzine E10 91 / Super Plus 98',
            'Regular/Benzine 91 / Super 95 / Super E10 95',
            'Regular/Benzine 91 / Super 95 / Super E10 95 / Regular/Benzine E10 91',
            'Super 95 / Regular/Benzine 91 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98 / Super E10 95',
            'Super Plus E10 98',
            'Super 95 / Regular/Benzine 91 / Super E10 95 / Super Plus E10 98 / Super Plus 98 / Regular/Benzine E10 91',
            'Super Plus 98 / Super E10 95',
            'Super 95 / Super E10 95 / Regular/Benzine 91',
            'Super 95 / Regular/Benzine E10 91 / Super E10 95',
            'Super 95 / Biodiesel',
            'Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98 / Super E10 95 / Regular/Benzine 91',
            'Regular/Benzine 91 / Super Plus 98 / Regular/Benzine E10 91 / Super 95 / Super E10 95 / Super Plus E10 98',
            'Super 95 / Super E10 95 / Super Plus 98',
            'Super Plus 98 / Super 95 / Super E10 95',
            'Super Plus E10 98 / Super 95 / Super Plus 98 / Super E10 95',
            'Regular/Benzine 91 / Super 95 / Super Plus 98',
            'Super E10 95 / Regular/Benzine 91 / Super 95 / Super Plus 98 / Regular/Benzine E10 91 / Super Plus E10 98',
            'Regular/Benzine 91 / Super 95 / Super E10 95 / Super Plus E10 98 / Super Plus 98 / Regular/Benzine E10 91',
            'Super E10 95 / Super 95 / Super Plus 98'            
            ]
df['Fuel'].replace(gasoline, 'Gasoline',inplace = True)

In [79]:
df['Fuel'].replace(['CNG','LPG', 'Liquid petroleum gas (LPG)','CNG (Particulate Filter)', 'Domestic gas H','Biogas'], 'Gas',inplace = True)

In [80]:
df['Fuel'].replace(['Others (Particulate Filter)','Others'], 'Others',inplace = True)

In [81]:
df.Fuel.value_counts(dropna = False)

Gasoline             8545
Diesel               7299
Gas                    64
Others                  6
Electric/Gasoline       4
Electric                1
Name: Fuel, dtype: int64

<font color='orange'> ** Impressions on the column 'Fuel' </font>
<font color='light blue'>
    
a. This column includes fuel types of cars .<br>
b. The data is clean but must be grouped <br>
c. Could be compared with the data in Type2_Fuel column.<br>
d. After comparison, one of them could be dropped.
</font>

### 37. 'Consumption' 

In [82]:
df['Consumption'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                   1906
[[3.9 l/100 km (comb)], [4.1 l/100 km (city)], [3.7 l/100 km (country)]]               304
[[4.2 l/100 km (comb)], [5 l/100 km (city)], [3.7 l/100 km (country)]]                 276
[[5.4 l/100 km (comb)], [6.8 l/100 km (city)], [4.5 l/100 km (country)]]               257
[[3.8 l/100 km (comb)], [4.3 l/100 km (city)], [3.5 l/100 km (country)]]               253
                                                                                      ... 
[[3.6 l/100 km (comb)], [], [4.4 l/100 km (country)]]                                    1
[\n, 4.8 l/100 km (comb), \n, 5.6 l/100 km (city), \n, 4.3 l/100 km (country), \n]       1
[[7.6 l/100 km (comb)], [], []]                                                          1
[[5.6 l/100 km (comb)], [7.6 l/100 km (city)], [4.4 l/100 km (country)]]                 1
[\n, 4.7 l/100 km (comb), \n, \n, \n]                                                    1

In [83]:
df['Consumption'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                   1906
[[3.9 l/100 km (comb)], [4.1 l/100 km (city)], [3.7 l/100 km (country)]]               304
[[4.2 l/100 km (comb)], [5 l/100 km (city)], [3.7 l/100 km (country)]]                 276
[[5.4 l/100 km (comb)], [6.8 l/100 km (city)], [4.5 l/100 km (country)]]               257
[[3.8 l/100 km (comb)], [4.3 l/100 km (city)], [3.5 l/100 km (country)]]               253
                                                                                      ... 
[[3.6 l/100 km (comb)], [], [4.4 l/100 km (country)]]                                    1
[\n, 4.8 l/100 km (comb), \n, 5.6 l/100 km (city), \n, 4.3 l/100 km (country), \n]       1
[[7.6 l/100 km (comb)], [], []]                                                          1
[[5.6 l/100 km (comb)], [7.6 l/100 km (city)], [4.4 l/100 km (country)]]                 1
[\n, 4.7 l/100 km (comb), \n, \n, \n]                                                    1

In [84]:
def consume_combined(a):
    if type(a) == list:
        if len(a) >3:
            for i in a:
                if 'comb' in i:
                    return i
        else:
            return a[0]            
    
    else:
        return a
    
df['Consumption_combined'] = df['Consumption'].apply(consume_combined)

In [85]:
def cleaning_consumption(a):
    if type(a) == list:
        if len(a) > 0:
            b = re.findall("\d\.?\d?", a[0])
            return b[0]
        else:
            return np.nan
    elif type(a) == str:
        b = re.findall("\d\.?\d?",a)
        return b[0]        
    else:
        return a


In [86]:
def consume_city(a):
    if type(a) == list:
        if len(a) >3:
            for i in a:
                if 'city' in i:
                    return i
        else:
            return a[1]            
    
    else:
        return a
    
df['Consumption_city'] = df['Consumption'].apply(consume_city)

In [87]:
def consume_country(a):
    if type(a)== list:
        if len(a) >3:
            for i in a:
                if 'country' in i:
                    return i
        else:
            return a[2]
            
    
    else:
        return a
    
df['Consumption_country'] = df['Consumption'].apply(consume_country)

In [89]:

df['Consumption_combined'] = df['Consumption_combined'].apply(cleaning_consumption).astype('float')
df['Consumption_city'] = df['Consumption_city'].apply(cleaning_consumption).astype('float')
df['Consumption_country'] = df['Consumption_country'].apply(cleaning_consumption).astype('float')

df.drop("Consumption",axis=1,inplace=True)


KeyError: "['Consumption'] not found in axis"

### 38. 'CO2 Emission' 

In [92]:
df['CO2 Emission'] = df['CO2 Emission'].str[0].str.findall('\d+').str[0].astype('float')



In [93]:
df.rename(columns={'CO2 Emission' : 'CO2 Emission(g CO2/km)'}, inplace = True)

In [94]:
df['CO2 Emission(g CO2/km)'].value_counts(dropna= False)

NaN      2436
120.0     740
99.0      545
97.0      537
104.0     501
         ... 
193.0       1
239.0       1
51.0        1
45.0        1
253.0       1
Name: CO2 Emission(g CO2/km), Length: 120, dtype: int64

### 39. 'Emission Class'

In [95]:
def emission(a):
    if type(a) == list:
        if len(a[0]) == 0:
            return np.nan
        else:
            return a[0].replace('\n','')
    else:
        return a
    
df['Emission Class'] = df['Emission Class'].apply(emission)

In [96]:
df['Emission Class'].replace(['Euro 6','Euro 6d-TEMP','Euro 6d', 'Euro 6c'], 'Euro 6', inplace = True)

In [97]:
df['Emission Class'].value_counts(dropna = False)

Euro 6    12173
NaN        3628
Euro 5       78
Euro 4       40
Name: Emission Class, dtype: int64

<font color='orange'> ** Impressions on the column 'Emission Class' </font>
<font color='light blue'>
    
a. This column includes emission classes of cars .<br>
b. The data is clean but must be grouped and 3628 NaNs <br>
c. Could be compared with the data in other columns.<br>
d. After comparison, others could be dropped.
</font>

### 40. 'Comfort & Convenience' 

In [247]:
df['\nComfort & Convenience\n'].head()
df.rename(columns={'\nComfort & Convenience\n':'Comfort&Convenience'}, inplace = True)


KeyError: '\nComfort & Convenience\n'

In [248]:
df['Comfort&Convenience'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                                                                                                                                                                                                                                                                                                     920
[Air conditioning, Electrical side mirrors, Hill Holder, Power windows]                                                                                                                                                                                                                                                                                                 216
[Air conditioning, Electrical side mirrors, Power windows]                                                                                                                                                                                                                      

In [249]:
df['Comfort&Convenience'] = df['Comfort&Convenience'].astype('str').str.replace('[','').str.replace("]",'')

In [251]:
df["Comfort&Convenience"].str.get_dummies(sep=',').T

0      1      2      3      4      \
 'Air suspension'                           0      0      0      0      0   
 'Armrest'                                  1      0      0      1      1   
 'Automatic climate control'                1      1      0      0      1   
 'Auxiliary heating'                        0      0      0      1      0   
 'Cruise control'                           1      0      1      0      0   
 'Electric Starter'                         0      0      0      0      0   
 'Electric tailgate'                        0      0      0      0      0   
 'Electrical side mirrors'                  1      0      1      1      1   
 'Electrically adjustable seats'            0      0      0      0      0   
 'Electrically heated windshield'           0      0      0      0      0   
 'Heads-up display'                         0      0      0      1      0   
 'Heated steering wheel'                    0      0      0      0      0   
 'Hill Holder'                              1      1      1      1      1   
 'Keyless central door lock'                0      0      0      0      0   
 'Leather seats'                            0      0      0      0      0   
 'Leather steering wheel'                   1      1      1      1      1   
 'Light sensor'                             1      0      0      1      1   
 'Lumbar support'                           0      1      0      1      0   
 'Massage seats'                            0      0      0      0      0   
 'Multi-function steering wheel'            1      0      1      1      1   
 'Navigation system'                        1      0      1      1      0   
 'Panorama roof'                            0      0      0      0      0   
 'Park Distance Control'                    1      0      1      0      1   
 'Parking assist system camera'             0      0      0      0      0   
 'Parking assist system self-steering'      0      0      0      0      0   
 'Parking assist system sensors front'      0      0      1      0      0   
 'Parking assist system sensors rear'       1      1      1      0      1   
 'Power windows'                            1      1      1      1      1   
 'Rain sensor'                              1      0      0      1      1   
 'Seat heating'                             1      0      1      1      0   
 'Seat ventilation'                         0      0      0      0      0   
 'Split rear seats'                         0      0      0      1      0   
 'Start-stop system'                        1      1      1      1      1   
 'Sunroof'                                  0      0      0      0      0   
 'Tinted windows'                           0      1      0      0      0   
 'Wind deflector'                           0      0      0      0      0   
 'Windshield'                               0      0      0      0      0   
'Air conditioning'                          1      1      1      0      1   
'Air suspension'                            0      0      0      1      0   
'Armrest'                                   0      0      0      0      0   
'Auxiliary heating'                         0      0      0      0      0   
'Cruise control'                            0      0      0      0      0   
'Electrical side mirrors'                   0      0      0      0      0   
'Electrically adjustable seats'             0      0      0      0      0   
'Electrically heated windshield'            0      0      0      0      0   
'Heads-up display'                          0      0      0      0      0   
'Heated steering wheel'                     0      0      0      0      0   
'Hill Holder'                               0      0      0      0      0   
'Leather steering wheel'                    0      0      0      0      0   
'Light sensor'                              0      0      0      0      0   
'Multi-function steering wheel'             0      0      0      0      0   
'Navigation system'                 

### 41. 'Entertainment & Media' 

In [239]:

df.rename(columns = {'\nEntertainment & Media\n': 'Entertainment&Media'}, inplace = True)
df['Entertainment&Media'].head()

0    [Bluetooth, Hands-free equipment, On-board com...
1    [Bluetooth, Hands-free equipment, On-board com...
2                             [MP3, On-board computer]
3    [Bluetooth, CD player, Hands-free equipment, M...
4    [Bluetooth, CD player, Hands-free equipment, M...
Name: Entertainment&Media, dtype: object

In [245]:
df['Entertainment&Media'] = df['Entertainment&Media'].astype('str').str.replace('[','').str.replace("]",'')




In [246]:
df["Entertainment&Media"].str.get_dummies(sep=',')

'CD player'   'Digital radio'   'Hands-free equipment'   'MP3'  \
0                 0                 0                        1       0   
1                 0                 0                        1       0   
2                 0                 0                        0       0   
3                 1                 0                        1       1   
4                 1                 0                        1       1   
...             ...               ...                      ...     ...   
15914             0                 1                        1       0   
15915             0                 1                        1       0   
15916             0                 0                        1       0   
15917             0                 1                        0       0   
15918             0                 0                        0       0   

        'On-board computer'   'Radio'   'Sound system'   'Television'   'USB'  \
0                         1         1                0              0       0   
1                         1         1                1              0       0   
2                         1         0                0              0       0   
3                         1         1                1              0       1   
4                         1         1                0              0       1   
...                     ...       ...              ...            ...     ...   
15914                     1         1                0              0       1   
15915                     0         1                0              0       1   
15916                     1         1                0              0       0   
15917                     0         1                0              0       1   
15918                     0         0                0              0       0   

       'Bluetooth'  'CD player'  'Digital radio'  'Hands-free equipment'  \
0                1            0                0                       0   
1                1            0                0                       0   
2                0            0                0                       0   
3                1            0                0                       0   
4                1            0                0                       0   
...            ...          ...              ...                     ...   
15914            1            0                0                       0   
15915            1            0                0                       0   
15916            1            0                0                       0   
15917            1            0                0                       0   
15918            0            0                0                       0   

       'MP3'  'On-board computer'  'Radio'  'Sound system'  'USB'  nan  
0          0                    0        0               0      0    0  
1          0                    0        0               0      0    0  
2          1                    0        0               0      0    0  
3          0                    0        0               0      0    0  
4          0                    0        0               0      0    0  
...      ...                  ...      ...             ...    ...  ...  
15914      0                    0        0               0      0    0  
15915      0                    0        0               0      0    0  
15916      0                    0        0               0      0    0  
15917      0                    0        0               0      0    0  
15918      0                    0        0               0      1    0  

[15919 rows x 19 columns]

### 42, Extras

In [105]:
df['\nExtras\n'].head()
df.rename(columns = {'\nExtras\n': 'Extras'}, inplace = True)
df.columns

Index(['make_model', 'body_type', 'price', 'vat', 'km', 'registration',
       'prev_owner', 'hp', 'Type', 'Next Inspection', 'Inspection new',
       'Warranty(months)', 'Body Color', 'Paint Type', 'Body Color Original',
       'Upholstery', 'Nr. of Doors', 'Nr. of Seats', 'Model Code',
       'Gearing Type', 'Displacement', 'Cylinders', 'Weight(kg)',
       'Drive chain', 'Fuel', 'CO2 Emission(g CO2/km)', 'Emission Class',
       'Comfort&Convenience', 'Entertainment&Media', 'Extras',
       '\nSafety & Security\n', 'description', 'Emission Label', 'Gears',
       'Country version', 'Electricity consumption', 'Last Service Date',
       'Other Fuel Types', 'Availability', 'Last Timing Belt Service Date',
       'Consumption_combined', 'Consumption_city', 'Consumption_country'],
      dtype='object')

### 43.Safety & Security ???

In [106]:
df['\nSafety & Security\n'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                                                                                                                                                                                                                                                                                                                        982
[ABS, Central door lock, Daytime running lights, Driver-side airbag, Electronic stability control, Fog lights, Immobilizer, Isofix, Passenger-side airbag, Power steering, Side airbag, Tire pressure monitoring system, Traction control]                                                                                                                                                 538
[ABS, Central door lock, Daytime running lights, Driver-side airbag, Electronic stability control, Immobilizer, Isofix, Passenger-side airbag, Power steering, Side airbag, Tire pressure monitoring system, Traction control]            

### 44. 'description' 

In [107]:
df['\nSafety & Security\n'].head()
df.rename(columns= {'\nSafety & Security\n': 'Safety & Security'}, inplace = True)

In [108]:
df.columns

Index(['make_model', 'body_type', 'price', 'vat', 'km', 'registration',
       'prev_owner', 'hp', 'Type', 'Next Inspection', 'Inspection new',
       'Warranty(months)', 'Body Color', 'Paint Type', 'Body Color Original',
       'Upholstery', 'Nr. of Doors', 'Nr. of Seats', 'Model Code',
       'Gearing Type', 'Displacement', 'Cylinders', 'Weight(kg)',
       'Drive chain', 'Fuel', 'CO2 Emission(g CO2/km)', 'Emission Class',
       'Comfort&Convenience', 'Entertainment&Media', 'Extras',
       'Safety & Security', 'description', 'Emission Label', 'Gears',
       'Country version', 'Electricity consumption', 'Last Service Date',
       'Other Fuel Types', 'Availability', 'Last Timing Belt Service Date',
       'Consumption_combined', 'Consumption_city', 'Consumption_country'],
      dtype='object')

### 45. 'Emission Label' ???

In [109]:
def emission(a):
    if type(a)== list:
        if len(a[0]) == 0:
            return np.nan
        else:
            return a[0].replace('\n','')
    else:
        return a
    
df['Emission Label'] = df['Emission Label'].apply(emission)
df['Emission Label'].value_counts(dropna = False, normalize = True)

NaN               0.752183
4 (Green)         0.223192
1 (No sticker)    0.023934
5 (Blue)          0.000503
3 (Yellow)        0.000126
2 (Red)           0.000063
Name: Emission Label, dtype: float64

In [110]:
df.drop('Emission Label',axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Emission Label' </font>
<font color='light blue'>
    
a. This column includes emission label of cars which is used in Germany .<br>
b. There are so many NaNs. 75% <br>
c. Dropped
</font>

### 46 Gears

In [111]:
def gears(a):
    if type(a)== list:
        return a[0].replace('\n','')
    else:
        return a
df['Gears'] = df['Gears'].apply(gears)    

In [112]:
df['Gears'].value_counts(dropna = False)

6      5822
NaN    4712
5      3239
7      1908
8       224
9         6
3         2
4         2
1         2
50        1
2         1
Name: Gears, dtype: int64

In [113]:
df[df["Gears"]==50.0]=5.0

<font color='orange'> ** Impressions on the column 'Gears' </font>
<font color='light blue'>
    
a. This column includes gear level of transmissions .<br>
b. There are 4712 NaNs. <br>
c. Regression could be check to see the relation with target column
</font>

### 47. Country

In [114]:
def country(a):
    if type(a)== list:
        return a[0].replace('\n','')
    else:
        return a
df['Country version'] = df['Country version'].apply(country)  

In [115]:
df['Country version'].value_counts(dropna = False, normalize = True)

NaN               0.523463
Germany           0.282807
Italy             0.065205
European Union    0.031849
Netherlands       0.029148
Spain             0.020416
Belgium           0.019725
Austria           0.013066
Czech Republic    0.003267
Poland            0.003078
France            0.002387
Denmark           0.002073
Hungary           0.001759
Japan             0.000503
Croatia           0.000251
Slovakia          0.000251
Sweden            0.000188
Bulgaria          0.000126
Romania           0.000126
Slovenia          0.000063
Egypt             0.000063
Luxembourg        0.000063
Switzerland       0.000063
Serbia            0.000063
Name: Country version, dtype: float64

<font color='orange'> ** Impressions on the column 'Country version' </font>
<font color='light blue'>
    
a. This column includes countries of cars .<br>
b. There are 8333 NaNs, %52, could be dropped <br>
c. Regression could be check to see the relation with target column
</font>

### 48. 'Electricity consumption'

In [116]:
df['Electricity consumption'].value_counts(dropna = False, normalize = True)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                          0.991394
[\n0 kWh/100 km (comb)\n]    0.008606
Name: Electricity consumption, dtype: float64

In [117]:
df.drop("Electricity consumption",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Electricity Consumption' </font>
<font color='light blue'>
    
a. This column includes electricity consumption of cars .<br>
b. 99% of the data NaN,  and dropped 
</font>

### 49. 'Last Service Date'

In [118]:
def last_service(a):
    if type(a)== list:
        return a[0].replace('\n','')
    else:
        return a
df['Last Service Date'] = df['Last Service Date'].apply(country)  

In [119]:
df['Last Service Date'].value_counts(dropna = False, normalize = True)

NaN            0.964445
05/2019        0.003455
01/2019        0.003078
02/2019        0.002890
06/2019        0.002890
04/2019        0.002764
03/2019        0.002638
12/2018        0.001570
10/2018        0.001382
05/2018        0.001256
01/2018        0.001194
07/2018        0.001068
06/2018        0.001068
09/2018        0.000942
11/2018        0.000942
04/2018        0.000879
08/2018        0.000691
03/2018        0.000565
\n02/2019\n    0.000565
06/2017        0.000440
\n12/2018\n    0.000440
\n05/2019\n    0.000377
\n03/2019\n    0.000314
\n07/2018\n    0.000251
02/2018        0.000251
\n06/2018\n    0.000251
\n04/2019\n    0.000251
\n06/2019\n    0.000188
05/2017        0.000188
12/2017        0.000188
01/2017        0.000188
10/2017        0.000188
02/2017        0.000188
\n09/2018\n    0.000126
06/2016        0.000126
\n04/2018\n    0.000126
\n01/2019\n    0.000126
11/2017        0.000126
\n08/2018\n    0.000126
04/2016        0.000126
\n11/2018\n    0.000126
\n10/2018\n    0

In [120]:
df.drop("Last Service Date",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Last Service Date' </font>
<font color='light blue'>
    
a. This column includes last service dates of cars .<br>
b. 96% of the data is NaN, and dropped 
</font>

### 50.'Other Fuel Types'

In [121]:
df['Other Fuel Types'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN             15039
[[], [], []]      880
Name: Other Fuel Types, dtype: int64

In [122]:
df.drop("Other Fuel Types",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Other Fuel Types' </font>
<font color='light blue'>
    
a. Empty column, and dropped
</font>

### 51. 'Availability'

In [123]:
df['Availability'].value_counts(dropna = False)

NaN                              15284
\nin 90 days from ordering\n       196
\nin 120 days from ordering\n      182
\nin 1 day from ordering\n          51
\nin 3 days from ordering\n         35
\nin 5 days from ordering\n         35
\nin 180 days from ordering\n       24
\nin 14 days from ordering\n        24
\nin 7 days from ordering\n         20
\nin 150 days from ordering\n       18
\nin 2 days from ordering\n         16
\nin 60 days from ordering\n        13
\nin 42 days from ordering\n        10
\nin 21 days from ordering\n         8
\nin 4 days from ordering\n          2
\nin 6 days from ordering\n          1
Name: Availability, dtype: int64

In [124]:
df.drop("Availability",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Availability' </font>
<font color='light blue'>
    
a. Many of the rows are empty, and dropped
</font>

### 52. 'Last Timing Belt Service Date'

In [125]:
df['Last Timing Belt Service Date'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                          15903
[\n04/2016\n]                    2
[\n12/1900\n]                    1
[\n01/1970\n]                    1
[\n04/2019\n]                    1
[\n01/2018\n]                    1
[\n06/2019\n]                    1
[\n02/2018\n]                    1
[\n02/2019\n]                    1
[\n01/2019\n]                    1
[\n06/2017\n]                    1
[\n05/2018\n, \nEuro 6\n]        1
[\n09/2018\n]                    1
[\n05/2019\n]                    1
[\n01/1900\n]                    1
[\n07/2018\n]                    1
Name: Last Timing Belt Service Date, dtype: int64

In [126]:
df.drop("Last Timing Belt Service Date",axis=1,inplace=True)

<font color='orange'> ** Impressions on the column 'Last Belt Service' </font>
<font color='light blue'>
    
a. Many of the rows are empty, could be dropped
</font>

### 53. 'Available from'

In [127]:
df['Available from'].value_counts(dropna = False)

KeyError: 'Available from'

In [ ]:
df.drop("Available from",axis=1,inplace=True)

In [ ]:
df.columns

<font color='orange'> ** Impressions on the column 'Available From'' </font>
<font color='light blue'>
    
a. Many of the rows are empty, could be dropped
</font>

## Import Libraries and Data

In [ ]:
df_dummied.to_csv("final_scout_dummy.csv", index=False)

In [ ]:
df.to_csv("final_scout_not_dummy.csv", index=False)